File: packages/shared/ts/src/pagination/cursor.ts
export interface CursorData {
  id: string;
  value: string;
}

/**
 * Encodes a document ID and sort field value into a base64 cursor string
 */
export function encodeCursor(id: string, sortValue: string): string {
  const data: CursorData = { id, value: sortValue };
  return Buffer.from(JSON.stringify(data)).toString('base64url');
}

/**
 * Decodes a base64 cursor string back into document ID and sort field value
 */
export function decodeCursor(cursor: string): CursorData {
  if (!cursor) {
    throw new Error('Empty cursor');
  }

  try {
    const json = Buffer.from(cursor, 'base64url').toString('utf-8');
    const data: CursorData = JSON.parse(json);

    if (!data.id) {
      throw new Error('Cursor missing document ID');
    }

    return data;
  } catch (error) {
    if (error instanceof SyntaxError) {
      throw new Error('Invalid cursor data');
    }
    throw error;
  }
}

File: packages/shared/go/pagination/cursor.go
package pagination

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
)

// cursorData holds the internal cursor state
type cursorData struct {
	ID    string `json:"id"`
	Value string `json:"value"`
}

// EncodeCursor encodes a document ID and sort field value into a base64 cursor string
func EncodeCursor(id string, sortValue string) string {
	data := cursorData{
		ID:    id,
		Value: sortValue,
	}

	jsonBytes, err := json.Marshal(data)
	if err != nil {
		return ""
	}

	return base64.URLEncoding.EncodeToString(jsonBytes)
}

// DecodeCursor decodes a base64 cursor string back into document ID and sort field value
func DecodeCursor(cursor string) (id string, sortValue string, err error) {
	if cursor == "" {
		return "", "", fmt.Errorf("empty cursor")
	}

	jsonBytes, err := base64.URLEncoding.DecodeString(cursor)
	if err != nil {
		return "", "", fmt.Errorf("invalid cursor encoding: %w", err)
	}

	var data cursorData
	if err := json.Unmarshal(jsonBytes, &data); err != nil {
		return "", "", fmt.Errorf("invalid cursor data: %w", err)
	}

	if data.ID == "" {
		return "", "", fmt.Errorf("cursor missing document ID")
	}

	return data.ID, data.Value, nil
}
